STM32 RTC : bypass shadow registers

jeromecoutant opened the pull request:
### Description

RTC implementation has been reworked a little in order to gain reactivity and avoid some latency.

Major impact is LPTICKER feature.

Tests are OK on my side for each STM32 family and each tool chain.

@marcemmers @bulislaw @c1728p9 @ashok-rao 

### Pull request type

    [ ] Fix
    [x] Refactor
    [ ] New target
    [ ] Feature
    [ ] Breaking change

0xc0170 commented:
@jeromecoutant Needs rebase now. While rebasing, can you add extend the commit msg (which registers are shadow? Because some chips as I noticed do not have them as I read in the comment code - from changes not clear to me how is this actually addressing the issue
`to gain reactivity and avoid some latency` - how did you achieve these? ).

jeromecoutant committed:
STM32 RTC : bypass shadow registers

- RTC_SSR for the subseconds
- RTC_TR for the time
- RTC_DR for the date

These registers were accessed through shadow registers which are synchronized with PCLK1 (APB1 clock).
They are now accessed directly in order to avoid waiting for the synchronization duration.

cmonr commented:
@marcemmers @bulislaw @c1728p9 @ashok-rao PR has been updated. Would y'all mind re-reviewing?

c1728p9 commented on targets/TARGET_STM/rtc_api.c:
```diff
@@ -247,40 +276,49 @@ void rtc_write(time_t t)
     }
+
+#if DEVICE_LPTICKER && !MBED_CONF_TARGET_LPTICKER_LPTIM
+    while (Read_SubSeconds != (RTC->SSR)) {
```
Is it possible that SSR increments and you get stuck in this loop?

jeromecoutant replied:
SSR sub seconds register is reset to PREDIV_S value every second.
This prescaler value is set only during init, so I would say we can't get stuck...

cmonr commented:
Looks like this'll need a rebase before it can continue :/

jeromecoutant commented:
@cmonr Rebase is done

cmonr commented:
/morph build
